Update paper.md (#652)
FAIRmat-NFDI/pynxtools: paper/paper.md

# NeXus reader and annotator (read\_nexus) (Sandor)

> __read_nexus__ is a command line tool to annotate NeXus data files. The tool is registered by the python ecosystem in the PATH and so it is practicly made available for immediate command line use in the environemnt the pynxtools python module is installed. The available command line options are explained if the argument _--help_ is passed when _read\_nexus_ is launched. These allow selecting a NeXus file and either documenting all or part of its content. Note that with no specific NeXus file provided as an argument, the tool will use by default an example NeXus file shiped together with the tool.  

Reviewer: ```suggestion
__read_nexus__ is a command line tool to annotate NeXus data files. The tool is registered by the Python ecosystem in the PATH and so it is practicly made available for immediate command line use in the environment the pynxtools python module is installed. The available command line options are explained if the argument _--help_ is passed when _read\_nexus_ is launched. These options allow selecting a NeXus file and either documenting all or part of its content. Note that with no specific NeXus file provided as an argument, the tool will use by default an example NeXus file shipped together with the package.
```

Reviewer: Not sure if detailed information on how it is available as a CLI tools is needed.

Could just say "__read_nexus__ is a tool to annotate NeXus data files, avaible from the command line if the pynxtools python module is installed"

Reviewer: Maybe the help part and the following sentence can be combined 
"The available command line options (explained using the argument _--help_) allow selecting a NeXus file and either documenting all or part of its content."

> _annotations -_ As it is expained above NeXus defines a huge list of concepts, the so called NeXus Vocabulary (https://manual.nexusformat.org/classes/index.html#classes-vocabulary-downloads). These include _Attributes_ and _Fields_ being organised in _Groups_ and their subgroups. NeXus files store their data items following the hierarchy of these groups (see details in https://manual.nexusformat.org/datarules.html). During reading a NeXus file, the proper interpretation of each data items requries following its group path and connecting it to the concept of the specific vocabulary item with the coresponding definition path. _read\_nexus_ traverse the input NeXus file and does it for each data items in it. Additionally to identifying the corresponding NeXus conpect, it immediately prints its definition. Since NeXus conepts follow inheritance, _read\_nexus_ also prints the defintions of the superclass NeXus concepts all along the inheritance chain. Note that if a data item in a NeXus file does not belong to any concepts which is allowed by the NeXus standard), _read\_nexus_ prints the message 'NOT IN SCHEMA'.  

Reviewer: ```suggestion
_annotations -_ As it is expained above, NeXus defines a large amount of concepts, the so called NeXus Vocabulary (https://manual.nexusformat.org/classes/index.html#classes-vocabulary-downloads). These include _Attributes_ and _Fields_ being organized in _Groups_ and their subgroups. NeXus files store their data items following the hierarchy of these groups (see details in https://manual.nexusformat.org/datarules.html). _read\_nexus_ traverses the input NeXus file and interprets each data item by following its group path and connecting it to the concept of the specific vocabulary item. It logs the concept's definition and, since NeXus concepts follow inheritance, those of its superclass concepts along the inheritance chain. If a data item in a NeXus file does not belong to any concepts (which is allowed by the NeXus standard), _read\_nexus_ logs the message 'NOT IN SCHEMA'.  
```

Suggestion to make it a bit more concise, probably even more cutting will be needed at some point.

> _default plotable -_ NeXus allows any data provider to specify which plottable data shall be presented by default when a given NeXus file is investigated (see details in https://manual.nexusformat.org/datarules.html#find-plottable-data). After traversing the whole file and visiting all data items, _read\_nexus_ also gives a summary of the default polottable by presenting its path and also its plotting details as specified by NXdata (https://manual.nexusformat.org/classes/base_classes/NXdata.html).  

Reviewer: ```suggestion
_default plottable -_ NeXus allows to specify which plottable data shall be presented by default for a given NeXus file (see details in https://manual.nexusformat.org/datarules.html#find-plottable-data). After traversing the whole file and visiting all data items, _read\_nexus_ additionally logs a summary of the default plottable by presenting its path and also its plotting details as specified by the NXdata definition (https://manual.nexusformat.org/classes/base_classes/NXdata.html).  
```

> __special features__ supporting particular semantic use of data items in a NeXus file is also incorporated to _read\_nexus_  

Reviewer: ```suggestion
In addition to this general annotation, particular semantic use of data items in a NeXus file is also incorporated:
```

Author: Maybe:  
In addition to this general annotation, the semantic use of data items in a NeXus file is also facilitated:

> _--concept_ This feature allows the user to query the NeXus file using a specific NeXus concept from the full vocabulary. _read\_nexus_ prints then all data item paths that corresponds to the provided NeXus concept.  

Reviewer: ```suggestion
_--concept_: This command line queries the NeXus file using a specific NeXus vocabulary item, logging all data item paths that correspond to this concept.
```

> _--documentation_ Instead of traversing the whole file, this feature allows the user to annotate only a single data item at a specific path in the NeXus file. This is parctical when only a few specific data items need to be investigated, e.g. those selected by a _--concept_ query.   

Reviewer: ```suggestion
_--documentation_: Instead of traversing the whole file, this feature annotates only a single data item at a specific path in the NeXus file. This is practical when only a few specific data items need to be investigated, e.g. those selected by a _--concept_ query.   
```

> __semantic use__ _read\_nexus_ provides semantic meaning to the data items by automating the annotation process of connecting data items of a NeXus file to the NeXus concepts defined with their full semantic meaning. Data processing applications, can use _read\_nexus_ to find data items in a NeXus file based on the NeXus concepts they should use and work on. They can process data files without the need of hardcoding any data path which would lead to incompatoibilities.  

Reviewer: ```suggestion
__semantic use__: read_nexus automates annotation, connecting data items to NeXus concepts with semantic meaning.
Data processing applications can use _read\_nexus_ to find data items in a NeXus file based on the NeXus concepts, without the need of hardcoding any data path which would lead to incompatibilities.  
```

I find this last sentence a bit hard to understand. Which incompabilities would arise from hardcoding?

> Also note that NeXus Ontology (https://github.com/nexusformat/NeXusOntology) ensures that all defined concepts of the NeXus vocabulary and their relationships including their defined restrictions are fully expressed using the semantic web technology standard OWL (https://www.w3.org/OWL/). Hence, _read\_nexus_ can be used to automatically connect data from any NeXus files to the semantic web.

Reviewer: I find this reference to the NeXus Ontology a bit misplaced here. We are not shipping any support for it with pynxtools (yet).

Reviewer 2: Agreed, it should be removed, also as the "automatically connect" is currently only possible because a human can e.g. use read_nexus to learn how to build a mapping rule from one concept to another but that is currently neither implemented via an automation nor do we have an example for it.

The NeXusOntology needs to find a prominent place in the NeXus paper though, agreed

Reviewer 2: agree with @lukaspie that we should check if there is enough space for listing all options, maybe priority rank them and report only the n most important functionalities and point to the link/doi with a sentence "further ... is available ... here"

Maybe the phrase "read_nexus offers querying the semantics of concepts with different granularities (all concepts in the file, a specific concept, etc.)

# NeXus reader and annotator (read\_nexus) (Sandor)

> __read_nexus__ is a command line tool to annotate NeXus data files. The tool is registered by the Python ecosystem in the PATH and so it is practicly made available for immediate command line use in the environment the pynxtools python module is installed. The available command line options are explained if the argument _--help_ is passed when _read\_nexus_ is launched. These options allow selecting a NeXus file and either documenting all or part of its content. Note that with no specific NeXus file provided as an argument, the tool will use by default an example NeXus file shipped together with the package.

Reviewer 2: The third sentence can also be shortened:
I thought that each command line tool (best practice) should offer --help, do we need to mention it, essentially this just states we follow best practices?

> _annotations -_ As it is expained above, NeXus defines a large amount of concepts, the so called NeXus Vocabulary (https://manual.nexusformat.org/classes/index.html#classes-vocabulary-downloads). These include _Attributes_ and _Fields_ being organized in _Groups_ and their subgroups. NeXus files store their data items following the hierarchy of these groups (see details in https://manual.nexusformat.org/datarules.html). _read\_nexus_ traverses the input NeXus file and interprets each data item by following its group path and connecting it to the concept of the specific vocabulary item. It logs the concept's definition and, since NeXus concepts follow inheritance, those of its superclass concepts along the inheritance chain. If a data item in a NeXus file does not belong to any concepts (which is allowed by the NeXus standard), _read\_nexus_ logs the message 'NOT IN SCHEMA'.  

Reviewer 2: above (better specific reference to paragraph)
instead of "large amount" give "..., NeXus defines .... concepts" give exact number and reference to the commit where you counted the number in.

Reviewer 2: Reformulate "The tool reports the definition of each concept" make a full stop definately before the "and, since", this and, since breaks the reading flow and is a related but not the same idea so ok to have a new sentence

> _default plottable -_ NeXus allows to specify which plottable data shall be presented by default for a given NeXus file (see details in https://manual.nexusformat.org/datarules.html#find-plottable-data). After traversing the whole file and visiting all data items, _read\_nexus_ additionally logs a summary of the default plottable by presenting its path and also its plotting details as specified by the NXdata definition (https://manual.nexusformat.org/classes/base_classes/NXdata.html).  

Reviewer 2: the verb "logs" is correct but ackward, maybe use "reports"

Reviewer 2: All these links are not permalinks, we should backup the definitions also on e.g. Zenodo to get a doi for the content.

Author: Suggestion taken. In case of links, there is another point. Let us harmonize them when all merged together.

> __special features__ supporting particular semantic use of data items in a NeXus file is also incorporated to _read\_nexus_  

Reviewer 2: "to" -> "into"

Author: rephrased
